src/hyperloglog.rs

```diff
@@ -359,12 +362,102 @@ where
     }
 }
 
+/// Serde support for `pdatastructs::hyperloglog::HyperLogLog`
+///
+/// # Examples
+/// ```
+/// use pdatastructs::hyperloglog::HyperLogLog;
+/// use pdatastructs::hyperloglog::serde::HyperLogLogExport;
+/// use serde_json;
+/// use std::hash::BuildHasherDefault;
+/// use std::collections::hash_map::DefaultHasher;
+///
+/// let mut hll = HyperLogLog::new(4);
+/// hll.add(&"some string");
+///
+/// let hll_export = HyperLogLogExport::from(&hll);
+/// let json = serde_json::to_string(&hll_export).expect("valid json");
+///
+/// // elsewhere
+/// let hll_export = serde_json::from_str(&json).expect("valid export");
+/// let hll: HyperLogLog<str, BuildHasherDefault<DefaultHasher>> = HyperLogLog::from_exported_hyperloglog(hll_export, BuildHasherDefault::default());
+///
+/// assert!(hll.count() > 0);
+/// ```
+#[cfg(feature = "serde")]
+pub mod serde {
+
+    use super::HyperLogLog;
+    use serde::{Deserialize, Serialize};
+    use std::fmt;
+    use std::hash::{BuildHasher, Hash};
+    use std::marker::PhantomData;
+
+    impl<T, B> HyperLogLog<T, B>
+    where
+        T: Hash + ?Sized,
+        B: BuildHasher + Clone + Eq,
+    {
+        /// Reconstruct HyperLogLog from a previously exported HyperLogLog
+        pub fn from_exported_hyperloglog(exported: HyperLogLogExport, buildhasher: B) -> Self {
+            Self {
+                registers: exported.registers,
+                b: exported.b,
+                buildhasher,
+                phantom: PhantomData,
+            }
+        }
+    }
+
+    /// Captures the underlying HyperLogLog data such that it can be serialized and deserialized
+    /// using serde
+    #[derive(Clone, Serialize, Deserialize, Eq, PartialEq)]
+    pub struct HyperLogLogExport {
```

Comment:
Instead of going through a helper struct, we should implement `Serialize` and `Deserialize` manually. In fact I think you don't need `serde-derive` here.

Regarding the hasher: either you need to constraint the `Serialize` impl to a hasher w/o any state or you also need to serialize the hasher. I think the latter one is technically the only proper way. That's important because the stdlib default hasher HAS state (the seed) and you shouldn't just drop that to the floor.

Reply:
Hey, thanks for the review. I've implemented this in terms of Serde manually now, but I don't think this is the right approach. 

I see your point and it is a potential problem if the hashers diverge, but I would simply document it and leave it to the users to deal with. None of the hashers in the `std` or other crates I've seen typically implement Serde, so then you're stuck with the orphan rule. Many hashes are also often not stable across architectures, or even time, so it's generally not a good idea storing these.

With this new implementation I as a user can't opt out of serializing the hasher, wasting bandwidth, I also can't serialize the HLL to formats not supporting serde, like protobuf.

In my view it's much more preferable to either have this auxiliary struct, or simply just mark the `registers` and `b` as `pub` and call it a day.